devito/petsc/iet/passes.py

```diff
@@ -0,0 +1,366 @@
+                  vec_replace_array, ksp_set_tols, ksp_set_type, ksp_get_pc,
+                  pc_set_type, ksp_set_from_ops])
+
+
+def create_matvec_callback(target, body, solver_objs, objs, struct):
+
+    # There will be 2 PETScArrays within the body
+    petsc_arrays = [i.function for i in FindSymbols('indexedbases').visit(body)
+                    if isinstance(i.function, PETScArray)]
+
+    # There will only be one PETScArray that is written to within this body and
+    # one PETScArray which corresponds to the 'seed' vector
+    petsc_arr_write, = FindSymbols('writes').visit(body)
+    petsc_arr_seed, = [i for i in petsc_arrays if i.function != petsc_arr_write.function]
+
+    # Struct needs to be defined explicitly here since CompositeObjects
+    # do not have 'liveness'
+    defn_struct = Definition(struct)
+
+    mat_get_dm = Call(petsc_call, [Call('MatGetDM', arguments=[
+        solver_objs['Jac'], Byref(objs['da'])])])
+
+    dm_get_app_context = Call(petsc_call, [Call('DMGetApplicationContext', arguments=[
+        objs['da'], Byref(struct._C_symbol)])])
+
+    dm_get_local_xvec = Call(petsc_call, [Call('DMGetLocalVector', arguments=[
+        objs['da'], Byref(solver_objs['X_local'])])])
+
+    global_to_local_begin = Call(petsc_call, [Call('DMGlobalToLocalBegin', arguments=[
+        objs['da'], solver_objs['X_global'], 'INSERT_VALUES', solver_objs['X_local']])])
+
+    global_to_local_end = Call(petsc_call, [Call('DMGlobalToLocalEnd', arguments=[
+        objs['da'], solver_objs['X_global'], 'INSERT_VALUES', solver_objs['X_local']])])
+
+    dm_get_local_yvec = Call(petsc_call, [Call('DMGetLocalVector', arguments=[
+        objs['da'], Byref(solver_objs['Y_local'])])])
+
+    vec_get_array_y = Call(petsc_call, [Call('VecGetArray', arguments=[
+        solver_objs['Y_local'], Byref(petsc_arr_write.function)])])
+
+    vec_get_array_x = Call(petsc_call, [Call('VecGetArray', arguments=[
+        solver_objs['X_local'], Byref(petsc_arrays[0])])])
+
+    dm_get_local_info = Call(petsc_call, [Call('DMDAGetLocalInfo', arguments=[
+        objs['da'], Byref(petsc_arrays[0].function.dmda_info)])])
+
+    casts = [PointerCast(i.function) for i in petsc_arrays]
+
+    vec_restore_array_y = Call(petsc_call, [Call('VecRestoreArray', arguments=[
+        solver_objs['Y_local'], Byref(petsc_arr_write)])])
+
+    vec_restore_array_x = Call(petsc_call, [Call('VecRestoreArray', arguments=[
+        solver_objs['X_local'], Byref(petsc_arr_seed)])])
+
+    dm_local_to_global_begin = Call(petsc_call, [Call('DMLocalToGlobalBegin', arguments=[
+        objs['da'], solver_objs['Y_local'], 'INSERT_VALUES', solver_objs['Y_global']])])
+
+    dm_local_to_global_end = Call(petsc_call, [Call('DMLocalToGlobalEnd', arguments=[
+        objs['da'], solver_objs['Y_local'], 'INSERT_VALUES', solver_objs['Y_global']])])
+
+    func_return = Call('PetscFunctionReturn', arguments=[0])
+
+    matvec_body = List(body=[
```

Comment:
Might be more readable if you break the `body` list into shorter lists with some suitable grouping which then get joined.

Reply:
yes, that or one word per line, it doesn't matter if it gets long